extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRuntimeClient.java

```diff
@@ -0,0 +1,238 @@
+
+        final OAuth2TokenImpl token = new OAuth2TokenImpl(auth, new JsonObject());
+
+        // always get the last token
+        token.principal().put("refresh_token", refreshToken);
+
+        token.refresh(new Handler<AsyncResult<Void>>() {
+            @Override
+            public void handle(AsyncResult<Void> result) {
+                if (result.succeeded()) {
+                    final String opaqueIdToken = token.opaqueIdToken();
+                    final String opaqueAccessToken = token.opaqueAccessToken();
+                    final String opaqueRefreshToken = token.opaqueRefreshToken() == null ? refreshToken
+                            : token.opaqueRefreshToken();
+
+                    AuthorizationCodeTokens tokens = new AuthorizationCodeTokens(opaqueIdToken, opaqueAccessToken,
+                            opaqueRefreshToken);
+                    emitter.complete(tokens);
+                } else {
+                    emitter.fail(result.cause());
+                }
+            }
+        });
+    }
+
+    public void getCodeFlowTokens(UniEmitter<? super AuthorizationCodeTokens> emitter, JsonObject params) {
+
+        auth.authenticate(params, new Handler<AsyncResult<User>>() {
+            @Override
+            public void handle(AsyncResult<User> result) {
+                if (result.succeeded()) {
+                    final AccessToken token = AccessToken.class.cast(result.result());
+                    final String opaqueIdToken = token.opaqueIdToken();
+                    final String opaqueAccessToken = token.opaqueAccessToken();
+                    final String opaqueRefreshToken = token.opaqueRefreshToken();
+
+                    AuthorizationCodeTokens tokens = new AuthorizationCodeTokens(opaqueIdToken, opaqueAccessToken,
+                            opaqueRefreshToken);
+                    emitter.complete(tokens);
+                } else {
+                    emitter.fail(result.cause());
+                }
+            }
+        });
+    }
+
+    public String getLogoutPath() {
+        return OAuth2AuthProviderImpl.class.cast(auth).getConfig().getLogoutPath();
+    }
+
+    public void decodeToken(String token, Handler<AsyncResult<AccessToken>> resultHandler) {
+        auth.decodeToken(token, resultHandler);
+    }
+
+    public void createUserInfoToken(UniEmitter<? super JsonObject> uniEmitter, RoutingContext vertxContext,
+            TokenAuthenticationRequest request) {
+        OAuth2TokenImpl tokenImpl = new OAuth2TokenImpl(auth, new JsonObject());
+        String accessToken = vertxContext.get("access_token");
+        if (accessToken == null) {
+            accessToken = request.getToken().getToken();
+        }
+        tokenImpl.principal().put("access_token", accessToken);
+        tokenImpl.userInfo(new Handler<AsyncResult<JsonObject>>() {
+            @Override
+            public void handle(AsyncResult<JsonObject> event) {
+                if (event.failed()) {
```

Comment:
There are plenty of async result <-> emitter code in this class. We should extract that into a specific method and maybe optimize (emitter is not great for this).